### PR TITLE
2+ Views (areas) share the same Model (document)

### PR DESCRIPTION
**Pull request by JordanMartinez**
This is the same result of the #231 but with a cleaner commit history.

### CODE REVIEW

File: richtextfx/src/main/java/org/fxmisc/richtext/StyledTextArea.java [hunk 1]
```diff
@@ -446,26 +449,41 @@ public void setStyleCodecs(Codec<S> textStyleCodec, Codec<PS> paragraphStyleCode
      * content model
      */
     private final EditableStyledDocument<S, PS> content;
+    protected final EditableStyledDocument<S, PS> getCloneDocument() {
```
Comment (TomasMikula) [inline richtextfx/src/main/java/org/fxmisc/richtext/StyledTextArea.java]: I would probably just call it `getDocument()`.

**Comment by TomasMikula** [comment]
`addSuspendable`, `removeSuspendable` looks a lot like adding/removing special kind of listeners. This suggests a question whether we can reuse some existing listener mechanism. We can! We can have a `SuspendableNo beingUpdatedProperty()` on `EditableStyledDocument` and add a listener that suspends the `StyledTextArea`'s `omniSuspendable`. I added [`Suspendable.suspendWhen(condition)`](https://github.com/TomasMikula/ReactFX/blob/da3f7f36aa4833ba864047d914604603e19b6a62/reactfx/src/main/java/org/reactfx/Suspendable.java#L86) to ReactFX to make it easier, so that one can now do something like

```java
manageSubscription(omniSuspendable.suspendWhen(content.beingUpdatedProperty()))
```

File: richtextfx/src/main/java/org/fxmisc/richtext/StyledTextArea.java (hunk 1 above)
Comment (JordanMartinez) [inline richtextfx/src/main/java/org/fxmisc/richtext/StyledTextArea.java]: I would but there's already a method with that name (see Line 336).

**Comment by JordanMartinez** [comment]
Aye... that would be much cleaner. Did you make a new release so I can update the code?

File: richtextfx/src/main/java/org/fxmisc/richtext/StyledTextArea.java (hunk 1 above)
Comment (TomasMikula) [inline richtextfx/src/main/java/org/fxmisc/richtext/StyledTextArea.java]: Right. What about `getContent()`?

**Comment by TomasMikula** [comment]
Yes, I updated the snapshot release (`2.0-SNAPSHOT`).

**Comment by JordanMartinez** [comment]
Great. I've had my first run-through it. If `EditableStyledDocument` has a `beingUpdatedProperty()` now, should I remove the `StyledTextArea`'s `beingUpdated` SuspendableNo ? (I know we ported it to ESD in the previous PR's exploration, but I reported it back to STA because it didn't add any value at that time.)

**Comment by JordanMartinez** [comment]
I ported `beingUpdated` to ESD. How's that?

File: richtextfx/src/main/java/org/fxmisc/richtext/StyledTextArea.java (hunk 1 above)
Comment (JordanMartinez) [inline richtextfx/src/main/java/org/fxmisc/richtext/StyledTextArea.java]: Sounds good. I've updated the method name.

**Comment by TomasMikula** [comment]
> Great. I've had my first run-through it. If EditableStyledDocument has a beingUpdatedProperty() now, should I remove the StyledTextArea's beingUpdated SuspendableNo ? (I know we ported it to ESD in the previous PR's exploration, but I reported it back to STA because it didn't add any value at that time.)

No, we need to keep another `beingUpdated` in STA, in order to guarantee the property that no change can be observed while `beingUpdated` is `false`. This way, the user can be sure that when `beingUpdated` changes from `true` to `false`, all other changes have been propagated and are thus consistent with each other (e.g. caret position, current paragraph, length, ...). Note that the order in which registered listeners are invoked is undefined (neither in JavaFX nor ReactFX). If STA observes the same `beingUpdated` as the user, the user could get the notification first, before STA releases the pending notifications.

**Comment by JordanMartinez** [comment]
Ah... I see. Fixed it.

**Comment by TomasMikula** [comment]
:+1:

**Comment by JordanMartinez** [comment]
Cool :smiley:

**Comment by TomasMikula** [comment]
So I merged this, but the supported use is limited. For example, if the document is edited from both areas, undo will not work very well. One area's undo will be regarded as a new edit in the other area. It is a complex problem on its own, so for now I would just say that such a use is not supported.

**Comment by JordanMartinez** [comment]
Dang.... Yeah, that's quite a problem.

On another note, when will there be a new mainstream release (meaning, not a snapshot release)?

**Comment by JordanMartinez** [comment]
> if the document is edited from both areas, undo will not work very well. One area's undo will be regarded as a new edit in the other area. It is a complex problem on its own, so for now I would just say that such a use is not supported.

I'll open an issue for this as well, just so it's on file.

**Comment by TomasMikula** [comment]
I can make a new release as soon as

1. all known regressions introduced by these refactorings are resolved.
2. all SNAPSHOT-versioned dependencies are released :)

**Comment by JordanMartinez** [comment]
> all known regressions introduced by these refactorings are resolved.

What are the known regressions? Would this include the Undo-related problem mentioned above?

> all SNAPSHOT-versioned dependencies are released :)
:) What else needs to be done in Flowless and ReactFX before they get their releases?

**Comment by TomasMikula** [comment]
> What are the known regressions?

#210, #220, #234 

> Would this include the Undo-related problem mentioned above?

That's not really a regression, since that was not supported at all before.

>  What else needs to be done in Flowless and ReactFX before they get their releases?

They might be good to go.

**Comment by JordanMartinez** [comment]
I've also noticed a problem in the richtextfx's build.gradle file. It lists UndoFX as a dependency twice: the first, with its transitive dependencies overrules the second, one without transitive dependencies. I've had to comment out the first instance or else I get compile warnings.

**Comment by TomasMikula** [comment]
Yeah, the first one should have been deleted when the second one was added. This is a mistake. Anyway, the `transitive = false` options will be removed when all dependencies use the same version of ReactFX as RichTextFX.

**Pull request merged**